core/src/main/java/de/danoeh/antennapod/core/storage/ItemEnqueuePositionCalculator.java

```diff
@@ -0,0 +1,95 @@
+package de.danoeh.antennapod.core.storage;
+
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+
+import java.util.List;
+
+import de.danoeh.antennapod.core.feed.FeedItem;
+import de.danoeh.antennapod.core.feed.FeedMedia;
+import de.danoeh.antennapod.core.preferences.UserPreferences.EnqueueLocation;
+import de.danoeh.antennapod.core.util.playback.Playable;
+
+/**
+ * @see DBWriter#addQueueItem(Context, boolean, long...) it uses the class to determine
+ * the positions of the {@link FeedItem} in the queue.
+ */
+class ItemEnqueuePositionCalculator {
+
+    @NonNull
+    private final EnqueueLocation enqueueLocation;
+
+    @VisibleForTesting
+    DownloadStateProvider downloadStateProvider = DownloadRequester.getInstance();
+
+    public ItemEnqueuePositionCalculator(@NonNull EnqueueLocation enqueueLocation) {
+        this.enqueueLocation = enqueueLocation;
+    }
+
+    /**
+     * Determine the position (0-based) that the item(s) should be inserted to the named queue.
+     *
+     * @param curQueue           the queue to which the item is to be inserted
+     * @param currentPlaying     the currently playing media
+     */
+    public int calcPosition(@NonNull List<FeedItem> curQueue, @Nullable Playable currentPlaying) {
+        switch (enqueueLocation) {
+            case BACK:
+                return curQueue.size();
+            case FRONT:
+                // return NOT 0, so that when a list of items are inserted, the items inserted
+                // keep the same order. Returning 0 will reverse the order
+                return getPositionOfFirstNonDownloadingItem(0, curQueue);
```

Comment:
Are comment and method still needed, now that the position is incremented externally?

Reply:
It is still needed: to handle the case that users taps to download in succession, resulting in multiple `addQueueItem()` calls. 
The comment is updated to make it clearer.